adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala

```diff
@@ -1806,3 +1897,42 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     }
   }
 }
+
+/**
+ * Ensures that our data is read and written in the same order.
+ *
+ * @note listStatus does not guarantee order, so the simplest solution here
+ *       was to override the getSplits and sort after we get all the splits,
+ *       while disallowing splitting within a file. This guarantees that our
+ *       data is read as it was written. Sorting overhead will be low because
+ *       we are only sorting a number of elements equal to the number of
+ *       partitions written.
+ */
+class ADAMInputFormat[T] extends ParquetInputFormat[T] {
+
+  /**
+   * Gets the splits of all the files for the JobContext object.
+   *
+   * @param job The JobContext object.
+   */
+  override def getSplits(job: JobContext): java.util.List[InputSplit] = {
+
+    val splits = new ListBuffer[InputSplit]()
+    // iterate over all files (in no particular order)
+    for (file <- listStatus(job)) {
+      val path = file.getPath
+      val length = file.getLen
+
+      if (length != 0) {
+        val fs = path.getFileSystem(job.getConfiguration)
+        val blkLocations = fs.getFileBlockLocations(file, 0, length)
+
+        splits.add(new FileSplit(path, 0, length, blkLocations(0).getHosts))
```

Comment:
This is how HDFS does it. I assume that it chooses index 0 because that is where the original data was written, rather than a replication. I'll add a note here about this.